wasm-chain-integration/src/v1/trie/low_level.rs

```diff
@@ -0,0 +1,3164 @@
+            // if there is a value at this node, then we have found our prefix.
+            if node.value.is_some() {
+                return true;
+            }
+            if let Ok(idx) = node.children.binary_search_by_key(&Chunk::new(*k), |x| x.key()) {
+                let c = unsafe { node.children.get_unchecked(idx) };
+                node_idx = c.index();
+            } else {
+                return false;
+            }
+        }
+        // we found a node that either has a value, or has children. In the first case
+        // it matches the key exactly, so is a prefix of it. In the latter, a key
+        // extends this node.
+        true
+    }
+
+    /// Delete the given key from the map. That is, decrease the reference count
+    /// by 1, and if this is the last occurrence of the key remove it from the
+    /// map. Return whether the key was in the map.
+    pub fn delete(&mut self, key: &[u8]) -> bool {
+        let mut node_idx = if let Some(root) = self.root {
+            root
+        } else {
+            // empty tree
+            return false;
+        };
+        let mut stack = Vec::new();
+        for k in key {
+            let node = unsafe { self.nodes.get_unchecked(node_idx) };
+            if let Ok(idx) = node.children.binary_search_by_key(&Chunk::new(*k), |x| x.key()) {
+                let c = unsafe { node.children.get_unchecked(idx) };
+                stack.push((node_idx, idx));
+                node_idx = c.index();
+            } else {
+                return false;
+            }
+        }
+        let node = unsafe { self.nodes.get_unchecked_mut(node_idx) };
+        let have_removed = node.value.is_some();
+        match node.value {
+            Some(ref mut value) if value.get() > 1 => {
+                *value = unsafe { NonZeroU32::new_unchecked(value.get() - 1) };
+                return true;
+            }
+            _ => node.value = None,
+        }
+        // back up and delete subtrees if needed
+        if node.children.is_empty() {
+            self.nodes.remove(node_idx);
+            while let Some((node_idx, child_idx)) = stack.pop() {
+                let node = unsafe { self.nodes.get_unchecked_mut(node_idx) };
+                node.children.remove(child_idx);
+                if !node.children.is_empty() || node.value.is_some() {
+                    break;
+                } else {
+                    self.nodes.remove(node_idx);
+                }
+            }
+            // delete the root, if needed
+            if let Some(root) = self.root {
+                if !self.nodes.contains(root) {
+                    self.root = None;
+                }
+            }
+        }
+        have_removed
+    }
+}
+
+#[derive(Debug)]
+/// A link to a shared occurrence of a value V.
+/// This is used in this module to construct trees, allowing for sharing of
+/// values in trees and subtrees in case of the persistent tree.
+pub struct Link<V> {
+    link: Arc<RwLock<V>>,
```

Comment:
Not sure I understand why the atomic part is needed and also the rwlock. Will revisit later, once I'm all the way through.

Reply:
I have added this comment to attempt to explain why Arc and RwLock are needed.
```
/// This [Link] achieves the following properties
/// - it is cheap to clone
/// - it allows for inner mutability
/// - it is safe to use in a concurrent context.
///
/// The cheap cloning is necessary to have efficient persistent trees. The
/// [Link] is used to point to children, as well as values and in the persistent
/// tree, modifying the value at a given point means we need to create copies of
/// the spine of the tree up to the root. This involves cloning pointers to any
/// parts that have not been modified. Thus these have to be cheap, so we use an
/// [Arc].
///
/// Inner mutability is needed so that the persistent tree may be loaded from
/// disk, as well as written to disk. This is achieved in combination with
/// [CachedRef]. Instead of using an [RwLock] we could instead use a [Mutex]
/// which would achieve the same API. However based on benchmarks the RwLock has
/// negligible overhead in the case of a single reader, and thus the [RwLock]
/// seems the more natural choice since it allows concurrent reads of the tree.
///
/// Finally, the reference counting must be atomic since parts of the tree are
/// dropped concurrently. While all the operations of the smart contract
/// execution engine are sequential, multiple states may be derived from the
/// same state in different threads. Currently this happens when using
/// invokeContract, but in the future with parallel block execution it might
/// also happen during normal block processing.
/// Additionally, if the Haskell runtime is configured with the parallel garbage
/// collector then parts of the tree might be dropped concurrently. This also
/// requires atomic reference counting.
```

Reply:
👌